### PR TITLE
Add io/fs#FS Driver #471

### LINKED ISSUE #471: feat: embed, io/fs support
**Is your feature request related to a problem? Please describe.**

The next Go 1.16 will add official file system interface and file embedding.

https://github.com/golang/go/issues/41190
https://github.com/golang/go/tree/master/src/io/fs

https://github.com/golang/go/issues/41191
https://github.com/golang/go/tree/master/src/embed

It would be great if golang-migrate supports these as source.Driver.

It replaces statik, go-bindata, and pkger.
Of course, this does not mean that they cannot be used.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

We will also need to use the build tag properly.

### PR DESCRIPTION
**Pull request by johejo**
close #471

### CODE REVIEW

File: source/iofs/README.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# iofs
+
+Driver with file system interface (`io/fs#FS`) supported from Go 1.16.
+
+This Driver cannot be used with versions below Go 1.15.
```
Comment (dhui) [inline source/iofs/README.md:5]: Update to: `This Driver cannot be used with Go versions 1.15 and below`

File: source/iofs/iofs.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/golang-migrate/migrate/v4/source/httpfs"
+)
+
+func init() {
+	source.Register("iofs", &Iofs{})
+}
+
+// Iofs is a source driver for io/fs#FS.
+type Iofs struct {
+	httpfs.PartialDriver
```
Comment (dhui) [inline source/iofs/iofs.go:21]: A refactor may be in-order.
It makes sense to go the other direction. e.g. add an `iofs.PartialDriver` and have `httpfs` embed that, since `io/fs.File` [interface](https://tip.golang.org/pkg/io/fs/#File) is a subset of the `net/http.File` [interface](https://golang.org/pkg/net/http/#File)

File: source/iofs/iofs.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+// Open by url does not supported with Iofs.
+func (i *Iofs) Open(url string) (source.Driver, error) {
+	return nil, errors.New("iofs driver does not support open by url")
+}
+
+// WithInstance wraps io/fs#FS as source.Driver.
+func WithInstance(fsys fs.FS, path string) (source.Driver, error) {
+	var i Iofs
+	if err := i.Init(http.FS(fsys), path); err != nil {
```
Comment (dhui) [inline source/iofs/iofs.go:32]: Converting from `fs.FS` to `http.FileSystem` may fail later if the underlying struct doesn't implement `io.Seeker`. `Seek()` isn't used in the `httpfs` source driver right now, but if that changes, this would break.

See: https://tip.golang.org/src/net/http/fs.go
```golang
func (f ioFile) Seek(offset int64, whence int) (int64, error) {
	s, ok := f.file.(io.Seeker)
	if !ok {
		return 0, errMissingSeek
	}
	return s.Seek(offset, whence)
}
```

**Review by dhui — COMMENTED** [review]
Thanks for the PR!

You'll need to add an `internal/cli/build_iofs.go` file and update `Makefile` with the build constraint for the official CLI to support `iofs`.

File: source/errors.go [hunk 4]
```diff
@@ -1,15 +1,18 @@
 package source
 
-import "os"
-
 // ErrDuplicateMigration is an error type for reporting duplicate migration
 // files.
 type ErrDuplicateMigration struct {
 	Migration
-	os.FileInfo
+	FileInfo
```
Comment (dhui) [inline source/errors.go:7]: Why not use `os.FileInfo` or [fs.FileInfo](https://tip.golang.org/pkg/io/fs/#FileInfo)?

You can use [fs.DirEntry.Info()](https://tip.golang.org/pkg/io/fs/#DirEntry) to get the `fs.FileInfo`

**Review by dhui — CHANGES_REQUESTED** [review]
Thanks for updating this to not depend on httpfs!

File: source/iofs/iofs.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+		if e.IsDir() {
+			continue
+		}
+		m, err := source.DefaultParse(e.Name())
+		if err != nil {
+			continue
+		}
+		file, err := e.Info()
+		if err != nil {
+			continue
```
Comment (johejo) [inline source/iofs/iofs.go:60]: Is continue OK like any other error case?

**Comment by johejo** [comment]
Thanks for the great review.
I got a better understanding of io/fs.

File: source/iofs/README.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
+# iofs
+
+Driver with file system interface (`io/fs#FS`) supported from Go 1.16.
+
+This Driver cannot be used with Go versions 1.15 and below.
```
Comment (slimsag) [inline source/iofs/README.md:5]: suggestion: Reword this to something like just:

> Package iofs provides a Go 1.16+ [`io/fs`](https://tip.golang.org/pkg/io/fs/) driver which is compatible with Go 1.16+ [file `//embed` directives](https://tip.golang.org/pkg/embed/).

Side note: I think all the documentation in this file would be better placed into a package-level godoc documentation comment, so that on https://pkg.go.dev this package shows up with package docs and usage examples similar to [this](https://pkg.go.dev/github.com/fluhus/godoc-tricks#pkg-overview) and [this](https://pkg.go.dev/github.com/fluhus/godoc-tricks#pkg-examples)

Feel free to ignore if this is not helpful :)

File: source/iofs/iofs.go (hunk 5 above)
Comment (slimsag) [reply source/iofs/iofs.go:60]: According to https://tip.golang.org/pkg/io/fs/#DirEntry an error in this case may mean:

> If the file has been removed or renamed since the directory read, Info may return an error satisfying errors.Is(err, ErrNotExist).

But I think it can also just represent any other IO error, in e.g. the case the FS interface is backed by a real FS and not `embed`. I think ignoring the error and `continue`ing is wrong here, instead any IO error should be reported back to the caller of the function I think (also in case of `DefaultParse` error above)

File: source/iofs/README.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"embed"
+	"log"
+
+	"github.com/golang-migrate/migrate/v4"
+	_ "github.com/golang-migrate/migrate/v4/database/postgres"
+	"github.com/golang-migrate/migrate/v4/source/iofs"
+)
+
+//go:embed migrations
```
Comment (slimsag) [inline source/iofs/README.md:23]:
```suggestion
//go:embed migrations/*.sql
```

May be useful to describe this to avoid people from accidentally embedding their own `README.md`, `.gitignore`, `.DS_Store`, etc.

File: source/iofs/README.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+var fs embed.FS
+
+func main() {
+	d, err := iofs.WithInstance(fs, "migrations")
```
Comment (slimsag) [inline source/iofs/README.md:27]:
```suggestion
	d, err := iofs.WithInstance(fs, ".")
```

If you accept [my other suggestion above](https://github.com/golang-migrate/migrate/pull/472/files#r523459865) this would need to change to `.`

File: source/iofs/iofs.go (hunk 5 above)
Comment (dhui) [reply source/iofs/iofs.go:60]: Agreed that errors should be returned and not ignored. Ignored errors will make bugs harder to debug.

File: source/iofs/iofs.go [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+	var i IoFS
+	if err := i.Init(fsys, path); err != nil {
+		return nil, fmt.Errorf("failed to init driver with path %s: %w", path, err)
+	}
+	return &i, nil
+}
+
+// Init prepares not initialized IoFS instance to read migrations from a
+// fs.FS instance and a relative path.
+func (p *IoFS) Init(fsys fs.FS, path string) error {
```
Comment (dhui) [inline source/iofs/iofs.go:43]: What are your thoughts about moving most of this implementation to a `PartialDriver`, similar to `httpfs`? We'd also need to rename `WithInstance()` to `New()` and change`IoFS` to `driver`. That way other drivers can embed the partial driver and implement their own `Open()` receiver function.

**Review by dhui — CHANGES_REQUESTED** [review]
Thanks for addressing my feedback!

File: .golangci.yml [hunk 10]
```diff
@@ -1,6 +1,6 @@
 run:
   # timeout for analysis, e.g. 30s, 5m, default is 1m
-  timeout: 2m
+  timeout: 5m
```
Comment (dhui) [inline .golangci.yml:3]: Leave this at `2m` I'll re-run any CI jobs/tests that fail due to timeout issues.

File: source/iofs/example_test.go [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+	"github.com/golang-migrate/migrate/v4"
+	_ "github.com/golang-migrate/migrate/v4/database/postgres"
+	"github.com/golang-migrate/migrate/v4/source/iofs"
+)
+
+//go:embed testdata/migrations/*.sql
+var fs embed.FS
+
+func Example() {
```
Comment (dhui) [inline source/iofs/example_test.go:17]: Thanks for using executable examples!

File: source/file/file.go [hunk 12]
```diff
@@ -1,31 +1,24 @@
...
 func init() {
 	source.Register("file", &File{})
 }
 
-type File struct {
-	httpfs.PartialDriver
-	url  string
-	path string
-}
+type File = file
```
Comment (dhui) [inline source/file/file.go:15]: Is a type alias necessary? I think you can define `type struct File` in both the go 1.16 and non-1.16 files.

File: source/file/file_go116.go [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"os"
+
+	"github.com/golang-migrate/migrate/v4/source"
+	"github.com/golang-migrate/migrate/v4/source/iofs"
+)
+
+type file struct {
+	iofs.Driver
```
Comment (dhui) [inline source/file/file_go116.go:13]: Thanks for doing this!

File: source/iofs/iofs.go [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+	"io"
+	"io/fs"
+	"path"
+	"strconv"
+
+	"github.com/golang-migrate/migrate/v4/source"
+)
+
+// Driver is a source driver that wraps io/fs#FS.
+type Driver struct {
```
Comment (dhui) [inline source/iofs/iofs.go:17]: Don't export this. e.g. rename to `driver`

File: source/iofs/iofs.go [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+	migrations *source.Migrations
+	fsys       fs.FS
+	path       string
+}
+
+// NewDriver returns a new Driver from io/fs#FS and a relative path.
+func NewDriver(fsys fs.FS, path string) (source.Driver, error) {
```
Comment (dhui) [inline source/iofs/iofs.go:24]: Rename to `New` to follow the same pattern as [httpfs.New()](https://pkg.go.dev/github.com/golang-migrate/migrate/v4@v4.13.0/source/httpfs#New)

File: source/iofs/iofs.go [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+
+// Open is part of source.Driver interface implementation.
+// Open panics when called directly.
+func (d *Driver) Open(url string) (source.Driver, error) {
+	panic("iofs: Driver does not support open with url")
+}
+
+// Init prepares not initialized IoFS instance to read migrations from a
+// io/fs#FS instance and a relative path.
+func (d *Driver) Init(fsys fs.FS, path string) error {
```
Comment (dhui) [inline source/iofs/iofs.go:40]: Move all of the driver receiver methods (except for `Open()`) to `PartialDriver` to follow the same patter than [httpfs](https://pkg.go.dev/github.com/golang-migrate/migrate/v4@v4.13.0/source/httpfs#PartialDriver) uses.

**Review by dhui — CHANGES_REQUESTED** [review]

File: source/iofs/iofs.go (hunk 14 above)
Comment (johejo) [reply source/iofs/iofs.go:17]: If did not export it, you will not be able to embed it in the file driver.
httpfs.PartialDriver is public now.

Comment (johejo) [reply source/iofs/iofs.go:17]: I understand it, I will separate into `PartialDriver` and `driver`.

File: source/file/file.go (hunk 12 above)
Comment (johejo) [reply source/file/file.go:15]: What kind of struct do you envision?
Sorry, I can't understand well.

Comment (dhui) [reply source/file/file.go:15]: I *think* you can avoid the type alias by doing:
```golang
// In file_go115.go
type File struct {
	httpfs.PartialDriver
	url  string
	path string
}

// In file_go116.go
type File struct {
	iofs.PartialDriver
	url  string
	path string
}
```

File: source/iofs/iofs.go [hunk 17]
```diff
@@ -31,13 +28,25 @@ func NewDriver(fsys fs.FS, path string) (source.Driver, error) {
 
 // Open is part of source.Driver interface implementation.
 // Open panics when called directly.
-func (d *Driver) Open(url string) (source.Driver, error) {
-	panic("iofs: Driver does not support open with url")
+func (d *driver) Open(url string) (source.Driver, error) {
+	panic("iofs: driver does not support open with url")
```
Comment (dhui) [inline source/iofs/iofs.go:32]: Don't panic and instead return an error: `return nil, errors.New("Open() cannot be called on the iofs passthrough driver")`

**Review by dhui — CHANGES_REQUESTED** [review]
Thanks for making the `PartialDriver` change! One last thing and we're good to go!

**Review by dhui — APPROVED** [review]

File: source/iofs/iofs_test.go [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+	"embed"
+	"testing"
+
+	"github.com/golang-migrate/migrate/v4/source/iofs"
+	st "github.com/golang-migrate/migrate/v4/source/testing"
+)
+
+func Test(t *testing.T) {
+	//go:embed testdata/migrations/*.sql
+	var fs embed.FS
```
Comment (jeffwidman) [inline source/iofs/iofs_test.go:15]: I'm just curious--how did this even work? 

Per the [golang docs](https://golang.org/pkg/embed/#hdr-Directives):
> The //go:embed directive can be used with both exported and unexported variables, depending on whether the package wants to make the data available to other packages. It can only be used with global variables at package scope, not with local variables.

To me this looks like a local variable, not a global variable at package scope... What am I missing?

Comment (johejo) [reply source/iofs/iofs_test.go:15]: Local variables were still supported when implementing this PR.
https://github.com/golang/go/issues/43216
It is a remnant of that.

Comment (jeffwidman) [reply source/iofs/iofs_test.go:15]: Ah, thank you!
